### PR TITLE
Set up AI and qcthat

### LINKED ISSUE #114: Implement `{qcthat}` framework in `gsm.core`
## Task Summary

Implement the [qcthat](https://gilead-biostats.github.io/qcthat/) framework to tie issues to unit tests and qualification tests to automate a report for each issue and each milestone to ensure each new feature has associated tests.

## Problem / Rationale

Part of the larger effort to modernize our qualification strategy

## QC Approach

In addition to standard QC (e.g., code reviews, automated checks) the following QC measures will be implemented:

[ ] Qualification Test via Double Programming
[ ] Unit Tests
[ ] User Tests (e.g. visual comparison)

### PR DESCRIPTION
**Pull request by jonthegeek**
# THIS PR SHOULD BE REVIEWED VERY CAREFULLY

## Overview
Implemented qcthat workflows & labels (`qcthat::use_qcthat()`), set up AI basics, and bootstrapped qcthat tags via qcthat AI (slightly modified here as part of ongoing experimentation).

We definitely need to create some requirement issues for things that existed when this repo was created. The majority of tests are *not* linked to any issues.

## Test Notes/Sample Code
The qcthat workflow will fail until we do the patch release.

I tried to revert anything that wasn't directly related to this PR, by my air formatter auto-styled a bunch of things that aren't currently styled "properly". We should do that in a separate PR sometime, but I wanted it to be clear what was being tagged.

In theory this PR's pkgdown site should have a report about what got tagged at `{pkgdown_site_url}/test_tag_reasons.html`. We should delete that report (from `inst/pkgdown/assets`) before merging the PR.

## Connected Issues

- Closes #114

### CODE REVIEW

File: .github/workflows/qcthat.yaml [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+          )
+        run: |
+          issueTestMatrix <- readRDS("ITM.rds")
+          qcthat::LoadUATIssues()
+          qcthat::CommentAllReports(
+            dfITM = issueTestMatrix,
+            lglPR = as.logical("${{ env.qcthat_PR_REPORT }}"),
+            lglMilestone = as.logical("${{ env.qcthat_MILESTONE_REPORT }}"),
+            lglCompleted = as.logical("${{ env.qcthat_COMPLETED_REPORT }}"),
+            lglUAT = as.logical("${{ env.qcthat_UAT }}")
```
[the remark this answers is not in the document]
Comment (jonthegeek) [reply .github/workflows/qcthat.yaml:108]: Incorrect. `as.logical("true")` is `TRUE`.

File: .github/workflows/qcthat.yaml [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        run: |
+          Rscript -e "qcthat::TriggerUAT()"
+
+      - name: Generate Issue-Test Matrix
+        if: >-
+          (env.qcthat_PR_REPORT == 'true' || env.qcthat_RELEASE_REPORT == 'true' || env.qcthat_FAIL_FOR_TEST_FAILURES == 'true') && (
+            github.event_name == 'push' ||
+            github.event_name == 'pull_request' ||
+            github.event_name == 'release' ||
+            (github.event_name == 'workflow_dispatch' && inputs.issueNumber == '')
```
[the remark this answers is not in the document]
Comment (jonthegeek) [reply .github/workflows/qcthat.yaml:82]: Logged in qcthat.

File: .Rbuildignore [hunk 3]
```diff
@@ -15,3 +15,5 @@
 ^inst/snapshots$
 ^inst/demo$
 ^inst/README\.md$
+^\.positai$
+^AGENTS\.md$
```
[the remark this answers is not in the document]
Comment (jonthegeek) [reply .Rbuildignore:19]: Things in inst are safe plus we're deleting this before merge.

File: tests/testthat/test-RunQuery.R [hunk 4]
```diff
@@ -80,7 +80,7 @@ test_that("RunQuery checks if all templated columns are found in lMapping", {
   )
 })
 
-test_that("RunQuery applies schema appropriately", {
+test_that("RunQuery applies schema appropriately (#22, #45)", {
```
Comment (lauramaxwell) [inline tests/testthat/test-RunQuery.R:83]: these issues makes sense

File: tests/testthat/test-qual_T5_2.R [hunk 5]
```diff
@@ -17,7 +17,7 @@ partial_mapped_workflows <- map(
 )
 
 ## Test Code
-testthat::test_that("Qual: Given appropriate raw participant-level data, flag values are correctly assigned as NA for sites with low enrollment (#116)", {
+testthat::test_that("Qual: Given appropriate raw participant-level data, flag values are correctly assigned as NA for sites with low enrollment (#77, #116)", {
```
Comment (lauramaxwell) [inline tests/testthat/test-qual_T5_2.R]: #116 makes sense, but i don't think #77 is relevant

File: tests/testthat/_snaps/util-MakeWorkflowList.md [hunk 6]
```diff
@@ -52,7 +52,7 @@
       
       
 
-# invalid data returns list NULL elements
+# invalid data returns list NULL elements (#43)
```
Comment (lauramaxwell) [inline tests/testthat/_snaps/util-MakeWorkflowList.md:55]: There was no PR attached to #43, so it's interesting that it pulled this. the issue relates to MakeWorkflowList, but i don't think it is relevant to this test

File: tests/testthat/_snaps/util-checkSpec.md [hunk 7]
```diff
@@ -33,7 +33,7 @@
       Not all specified columns in the spec are present in the data, missing columns are: df1$b
       Not all specified columns in the spec are present in the data, missing columns are: df2$y
 
-# Validate column type works
+# Validate column type works (#22)
```
Comment (lauramaxwell) [inline tests/testthat/_snaps/util-checkSpec.md:36]: should we be adding issues to snapshots?

File: tests/testthat/test-Analyze_NormalApprox_PredictBounds.R [hunk 8]
```diff
@@ -45,7 +45,7 @@ test_that("Analyze_NormalApprox_PredictBounds handles missing vThreshold correct
   expect_equal(sort(unique(dfBounds$Threshold)), sort(c(-3, -2, 0, 2, 3)))
 })
 
-test_that("Analyze_NormalApprox_PredictBounds processes data correctly", {
+test_that("Analyze_NormalApprox_PredictBounds processes data correctly (#21)", {
```
Comment (lauramaxwell) [inline tests/testthat/test-Analyze_NormalApprox_PredictBounds.R:48]: This issue #21 does address some warnings in this function's tests, so i suppose it makes sense to link it, but it isn't the primary objective of this test

File: tests/testthat/test-RunQuery.R [hunk 9]
```diff
@@ -132,7 +132,7 @@ test_that("RunQuery applies schema appropriately", {
   expect_equal(class(result$Tenured), "logical")
 })
 
-test_that("RunQuery applies incomplete schema appropriately", {
+test_that("RunQuery applies incomplete schema appropriately (#70, #71)", {
```
Comment (lauramaxwell) [inline tests/testthat/test-RunQuery.R:135]: these issues make sense!

File: tests/testthat/test-RunQuery.R [hunk 10]
```diff
@@ -162,7 +162,7 @@ test_that("RunQuery applies incomplete schema appropriately", {
 })
 
 
-test_that("RunQuery parses invalid date/times correctly", {
+test_that("RunQuery parses invalid date/times correctly (#44)", {
```
Comment (lauramaxwell) [inline tests/testthat/test-RunQuery.R:165]: This issue makes sense!

File: tests/testthat/test-RunStep.R [hunk 11]
```diff
@@ -54,8 +54,11 @@ test_that("Passes direct value parameters correctly", {
   expect_equal(result$y, "100")
 })
 
-test_that("Passes direct value vector parameters correctly", {
-  lStep <- list(name = "dummy_function", params = list(x = "meta1", y = c(1, 2, 3)))
+test_that("Passes direct value vector parameters correctly (#23)", {
```
Comment (lauramaxwell) [inline tests/testthat/test-RunStep.R:57]: this issue makes sense!

File: tests/testthat/test-RunStep.R [hunk 12]
```diff
@@ -107,7 +110,7 @@ test_that("RunStep will run a function without a namespace", {
   expect_equal(result, head(Theoph))
 })
 
-test_that("RunStep will run a function with no parameters", {
+test_that("RunStep will run a function with no parameters (#31)", {
```
Comment (lauramaxwell) [inline tests/testthat/test-RunStep.R:113]: This issue makes sense!

File: tests/testthat/test-util-checkSpec.R [hunk 13]
```diff
@@ -88,7 +88,7 @@ test_that("Multiple missing columns are correctly reported", {
   expect_snapshot(CheckSpec(lData, lSpec))
 })
 
-test_that("Validate column type works", {
+test_that("Validate column type works (#22)", {
```
Comment (lauramaxwell) [inline tests/testthat/test-util-checkSpec.R:91]: This issue makes sense!

File: tests/testthat/test-util-MakeWorkflowList.R [hunk 14]
```diff
@@ -17,40 +17,38 @@ test_that("Metadata is returned as expected", {
...
-  {
-    bRecursive <- TRUE
-    ### strNames - testing strNames equal to random numeric array
-    expect_snapshot(
-      wf_list <- MakeWorkflowList(
-        strNames = "kri8675309",
-        strPath = test_path("testdata"),
-        bRecursive = bRecursive
-      )
+test_that("invalid data returns list NULL elements (#43)", {
```
Comment (lauramaxwell) [inline tests/testthat/test-util-MakeWorkflowList.R:20]: #43 issue sort of makes sense for lines 35-41 of this test

File: tests/testthat/test-Analyze_NormalApprox.R [hunk 15]
```diff
@@ -20,7 +20,7 @@ test_that("binary output created as expected and has correct structure", {
   expect_equal(unique(binary$GroupID), c("166", "86", "76"))
 })
 
-test_that("rate output created as expected and has correct structure", {
+test_that("rate output created as expected and has correct structure (#21)", {
```
Comment (lauramaxwell) [inline tests/testthat/test-Analyze_NormalApprox.R:23]: This issue #21 does address some warnings in this function's tests, so i suppose it makes sense to link it, but it isn't the primary objective of this test

**Review by lauramaxwell — COMMENTED** [review]
went over the tagged tests- mixed results, but not bad! Happy to take a look at the ones that don't match and find a better issue- let me know how you want to proceed

**Comment by jonthegeek** [comment]
> went over the tagged tests- mixed results, but not bad! Happy to take a look at the ones that don't match and find a better issue- let me know how you want to proceed

I wouldn't worry about adding replacements as part of this PR, but we should remove any that don't fit. The aim here is "0% wrong" not "100% covered". Humans will have to sort out coverage in a later round.

I have at least one more specific comment that I'll add in the individual comment threads.

File: tests/testthat/_snaps/util-MakeWorkflowList.md (hunk 6 above)
Comment (jonthegeek) [reply tests/testthat/_snaps/util-MakeWorkflowList.md:55]: This was almost definitely tagged purely by name, after the "when did this change?" round.

File: tests/testthat/_snaps/util-checkSpec.md (hunk 7 above)
Comment (jonthegeek) [reply tests/testthat/_snaps/util-checkSpec.md:36]: Snapshot titles are based on the PR descriptions (and thus they update when the tests update). We shouldn't review the `_snaps`; these were updated by re-running tests after the tags were applied.

File: tests/testthat/test-Analyze_NormalApprox.R (hunk 15 above)
Comment (jonthegeek) [reply tests/testthat/test-Analyze_NormalApprox.R:23]: Added in "bonus round" (doesn't use commit history). Probably fine as a bonus issue, but shouldn't be the ONLY issue. So I'd call this one basically neutral as for "Is this bonus process helpful?"

File: tests/testthat/test-Analyze_NormalApprox_PredictBounds.R (hunk 8 above)
Comment (jonthegeek) [reply tests/testthat/test-Analyze_NormalApprox_PredictBounds.R:48]: Added in "bonus round" (doesn't use commit history). Probably fine as a bonus issue, but shouldn't be the ONLY issue. So I'd call this one basically neutral as for "Is this bonus process helpful?"

File: tests/testthat/test-qual_T5_2.R (hunk 5 above)
Comment (jonthegeek) [reply tests/testthat/test-qual_T5_2.R]: #116 was evidently added manually. It added #77 in the "bonus round". Scoring this one as negative for "Is this bonus process helpful?"

File: tests/testthat/test-RunQuery.R (hunk 4 above)
Comment (jonthegeek) [reply tests/testthat/test-RunQuery.R:83]: Commit matches.

File: tests/testthat/test-RunQuery.R (hunk 9 above)
Comment (jonthegeek) [reply tests/testthat/test-RunQuery.R:135]: Commit matches.

File: tests/testthat/test-RunQuery.R (hunk 10 above)
Comment (jonthegeek) [reply tests/testthat/test-RunQuery.R:165]: Commit matches.

File: tests/testthat/test-RunStep.R (hunk 11 above)
Comment (jonthegeek) [reply tests/testthat/test-RunStep.R:57]: Commit matches.

File: tests/testthat/test-RunStep.R (hunk 12 above)
Comment (jonthegeek) [reply tests/testthat/test-RunStep.R:113]: Commit matches.

File: tests/testthat/test-util-checkSpec.R (hunk 13 above)
Comment (jonthegeek) [reply tests/testthat/test-util-checkSpec.R:91]: Commit matches.

File: tests/testthat/test-util-MakeWorkflowList.R (hunk 14 above)
Comment (jonthegeek) [reply tests/testthat/test-util-MakeWorkflowList.R:20]: This was added in the bonus round. I think this is a valid tag (just not the ONLY tag, ideally). Calling this a success for that round.

**Comment by jonthegeek** [comment]
Saving this report (auto-generated during tagging) in the PR as a comment so we don't need to keep the separate qmd/html:

This document records the reasoning behind each test–issue association added in this tagging pass. Tags were applied via `qcthat` commit-based matching (Steps 1–2), title matching (Step 3), and a manual bonus round targeting closed issues not yet covered by any test.

Each section corresponds to a test file. The **Reason** column notes the match type and the logic used.

---

## test-Analyze_NormalApprox.R

| Test | Issue | Reason |
|------|-------|--------|
| rate output created as expected and has correct structure | [#21](https://github.com/Gilead-BioStats/gsm.core/issues/21) | **Bonus round.** Issue #21 explicitly names `test-Analyze_NormalApprox.R` as one of the files with captured warnings. This test uses `quiet_Analyze_NormalApprox()`, a wrapper that suppresses expected messages/warnings — the direct implementation of the "capture warnings so real issues aren't missed" fix described in #21. The `quiet_*` helper is the clearest signal in the file. |

---

## test-Analyze_NormalApprox_PredictBounds.R

| Test | Issue | Reason |
|------|-------|--------|
| Analyze_NormalApprox_PredictBounds processes data correctly | [#21](https://github.com/Gilead-BioStats/gsm.core/issues/21) | **Bonus round.** Issue #21 explicitly names `test-Analyze_NormalApprox_PredictBounds.R`. This test uses `quiet_Analyze_NormalApprox_PredictBounds()`, the suppression wrapper that directly addresses the warning-capture concern in #21. |

---

## test-qual_T5_2.R

| Test | Issue | Reason |
|------|-------|--------|
| Qual: Given appropriate raw participant-level data, flag values are correctly assigned as NA for sites with low enrollment | [#77](https://github.com/Gilead-BioStats/gsm.core/issues/77) | **Bonus round.** Issue #77 is "Add `RiskScoreWeight` info to `Analysis_Flagged` via the `Flag()` function." The test calls `Flag(..., vRiskScoreWeight = c(32, 16, 0, 1, 2))` and then removes `Weight`/`WeightMax` columns for comparison — confirming these columns are expected to be present. The `vRiskScoreWeight` parameter usage is deliberate and tests the exact feature described in #77. Tag added alongside the pre-existing `#116`. |

---

## test-RunQuery.R

| Test | Issue | Reason |
|------|-------|--------|
| RunQuery applies schema appropriately | [#22](https://github.com/Gilead-BioStats/gsm.core/issues/22) | **Commit match.** Issue #22 is "Feature: DATETIME data type — support 'timestamp' format in RunQuery." The test includes `Birthtime = list(type = "timestamp")` in `lColumnMapping` and asserts `class(result$Birthtime) == c("POSIXct", "POSIXt")` — directly testing timestamp parsing. |
| RunQuery applies schema appropriately | [#45](https://github.com/Gilead-BioStats/gsm.core/issues/45) | **Commit match.** Issue #45 is "Feature: Support `logical` data type — map to BOOLEAN in duckdb." The test includes `Tenured = list(type = "logical")` and asserts `class(result$Tenured) == "logical"` — directly testing logical type coercion. |
| RunQuery applies incomplete schema appropriately | [#70](https://github.com/Gilead-BioStats/gsm.core/issues/70) | **Commit match.** Issue #70 is "Update `RunQuery` to not set `character` type if not specified." The test passes a `lColumnMapping` with one column that has no `type` field and asserts `class(result$emaN) == class(df$Name)` — verifying the original character type is preserved, not overridden with `character`. |
| RunQuery applies incomplete schema appropriately | [#71](https://github.com/Gilead-BioStats/gsm.core/issues/71) | **Commit match.** Issue #71 is "Bugfix: `RunQuery()` coerces undefined columns to `character`." Same test as above; #71 is the bug report underlying the #70 business requirement. Both describe the same behavioral fix tested here. |
| RunQuery parses invalid date/times correctly | [#44](https://github.com/Gilead-BioStats/gsm.core/issues/44) | **Commit match.** Issue #44 is "Bugfix: `RunQuery` throws an error on empty `Date` strings." The test passes `Birthday = c("1990JAN01", "1987-02-30", "")` and `Birthtime = c(...)` with invalid/empty values and asserts `all(is.na(result$Birthday))` and `all(is.na(result$Birthtime))` — exactly the `NA` behavior expected by #44. |

---

## test-RunStep.R

| Test | Issue | Reason |
|------|-------|--------|
| Passes direct value vector parameters correctly | [#23](https://github.com/Gilead-BioStats/gsm.core/issues/23) | **Commit match.** Issue #23 is "Bugfix: Message out of `RunStep` related to vector arguments should not be vectorized — one message should be emitted for a vector-type argument." The test uses `suppressMessages(expect_message(..., "y is of length 3"))` to confirm exactly one message is emitted for a 3-element vector, verifying the bug fix. |
| RunStep will run a function with no parameters | [#31](https://github.com/Gilead-BioStats/gsm.core/issues/31) | **Commit match.** Issue #31 is "Feature/Bug — allow no params in workflow steps. Functions that have all defaults or no parameters currently fail." The test calls `RunStep(list(name = "getwd"), list(), list())` with no `params` key and asserts the result equals `getwd()` — directly testing the fix. |

---

## test-util-checkSpec.R

| Test | Issue | Reason |
|------|-------|--------|
| Validate column type works | [#22](https://github.com/Gilead-BioStats/gsm.core/issues/22) | **Commit match.** Issue #22 adds support for `"timestamp"` as a recognized data type. The test spec includes `SnapshotDateTime = list(type = "timestamp")` and `expect_snapshot(CheckSpec(...))` to verify the spec check handles timestamp types correctly without error. |

---

## test-util-MakeWorkflowList.R

| Test | Issue | Reason |
|------|-------|--------|
| invalid data returns list NULL elements | [#43](https://github.com/Gilead-BioStats/gsm.core/issues/43) | **Bonus round.** Issue #43 is "`MakeWorkflowList` doesn't work with packages that aren't installed." The test explicitly calls `MakeWorkflowList(strPackage = "fake-pkg", ...)` and wraps it in `expect_error()`, verifying that a non-existent package now produces a clear error rather than a cryptic `strPath does not exist` message. |

---

## Skipped issues

The following closed issues were reviewed but no suitable test match was found:

| Issue | Title | Reason skipped |
|-------|-------|----------------|
| #1 | QC: Update ReadME | Documentation only |
| #2 | QC: Update vignettes | Documentation only |
| #7 | Feature: Add nMinNumerator to `Summarize()` | No test exercises `nMinNumerator`; `Summarize.R` shows `nMinDenominator` (deprecated) but not `nMinNumerator`. |
| #8 | Feature: Add PK data to `lSource` | Data file change only |
| #13 | QC: Update package data with PK updates | Data file change only |
| #19 | QC: Add unit test for new parameter `vColumns` to `Summarize` | `vColumns` is not present in the current `Summarize()` signature; no test clearly maps to it. |
| #24 | Bugfix: Fix outdated vignette features | Vignette only |
| #25 | Business Requirement: CRAN release of gsm.core | Too broad; no specific behavior tested |
| #28 | Business Requirement: Update Accrual Thresholds | YAML default changes; not tested in unit tests |
| #42 | Documentation: Update outdated links | Documentation only |
| #47 | QC: Add a new example incorporating the delta columns | Example/documentation only |
| #50 | Feature: Update `lSource` to comply with `gsm.mapping@fix-72` | Data/example change only |
| #56 | QC: update sample data | Data update only |
| #64 | Bugfix: `STUDCOMP` data bug from datasim | Data fix only |
| #66 | Call `Ingest` from `RunWorkflow` | RunWorkflow tests test happy-path outputs; none target Ingest specifically |
| #67 | Add `tryCatch` to `RunWorkflows` | No test for error-handling / tryCatch behavior in RunWorkflow |
| #77 (qual_T5_2) | See above — **tagged** | — |
| #79 | Update Contributor Guidelines | Documentation only |
| #80 | Add workflow tarball to releases | CI/CD only |
| #89 | Add site risk score metric to `reportingResults` | Issue body is sparse; no unit test found that specifically targets this metric |
| #91 | Fix links to images in vignettes | Documentation only |
| #96 | Add IE support in `.rda` files | Data file only |
| #101 | Refactor Examples | Example/documentation only |
| #104 | Update Readme.md | Documentation only |
| #105 | Incorporate `IE` into standard site-based KRI | No matching unit test found |
| #107 | Implement gsm.utils v0.2.0 workflows | Workflow folder; no package code tested |
| #112 | Update Quality Control section of Readme | Documentation only |
| #118 | Migrate qualification documentation | Documentation only |
| #119 | Update r-releaser workflow | CI/CD only |

**Comment by jonthegeek** [comment]
This is the skill that it used. We don't need this in the repo after this point, so I'm deleting it, but I want it logged:

---

```
---
name: tag-tests-with-issues
description: Identify likely GitHub issues connected to test cases. Use when asked to tag tests with issues or get started with qcthat.
---
```

# Tag tests with issues

Add issue references (e.g., `(#123)`) to test descriptions to connect tests to the features or bugs they address. Execute with minimal interaction; skip chat output intended only for user review.

Note: Files without a full path are in `tests/testthat`.

## Step 1: Load test and issue context

Extract all tests and process them file-by-file. `MapTestFilesToPotentialIssues()` can be slow, so pre-compute the issue-commit mappings once with `MapLongIssueCommits()` and pass the result to each call. Do not try to speed up the process by avoiding this function. Its output is required for `PrepareTestIssueContext()`, and the output of `PrepareTestIssueContext()` is your source of all context for the remaining steps.

```r
library(qcthat)

dfFileTests <- ExtractTestsFromFiles()
lFileTestsSplit <- split(dfFileTests, dfFileTests$File)

# Pre-compute issue-commit mappings once (avoids redundant API calls per file)
dfIssueCommitsLong <- MapLongIssueCommits()

dfPotentialIssues <- MapTestFilesToPotentialIssues(
  lFileTestsSplit[[1]],
  dfIssueCommitsLong = dfIssueCommitsLong
)
dfTestIssueContext <- PrepareTestIssueContext(dfPotentialIssues)
```

Process files one at a time (`lFileTestsSplit[[1]]`, then `lFileTestsSplit[[2]]`, etc.) unless told otherwise, but use the same `dfIssueCommitsLong` instance for each.

`PrepareTestIssueContext()` returns a data frame with columns:

| Column | Type | Description |
|--------|------|-------------|
| `Test` | character | Test description |
| `File` | character | Test file name |
| `LineStart` / `LineEnd` | integer | Line numbers in the test file |
| `Issues` | list of integer vectors | Issues already tagged in the description |
| `PotentialIssueDetails` | list of tibbles | One tibble per test with `Issue`, `Title`, `Body` for each potentially related issue, identified by matching commits that modified the test with commits that closed issues |
| `TestCode` | list of character vectors | The actual test code |

## Step 2: Match tests to issues

For each test, compare code and description against `PotentialIssueDetails`. Do not use `git blame`, `gh`, or other tools — `dfTestIssueContext` contains everything you need. Matches must come only from the `Issue` column of `PotentialIssueDetails` for that test. Most tests match one issue; some match zero; a few match more than one.

### Decision process

For every test with non-empty `PotentialIssueDetails`:

**1. Read `TestCode` to understand what the test actually does.** Descriptions can be vague — code is the ground truth. Extract: the primary function called, what `expect_*()` calls verify, and any edge cases or context.

**2. Read the ENTIRE `Body` of each potential issue.** Titles are often vague or misleading. Note whether the body names the function being tested, describes the specific behavior, and what the issue is *for*: implementing the feature, fixing a bug, or something unrelated.

**3. Look for a function name match.** If an issue body mentions the function AND is about implementing, fixing, or extending it, that is a strong match. The issue must be *about* the function — one that merely mentions it in passing (e.g., "this refactoring touched `ProcessPayment()`") is not a match.

**4. If no function name match, look for a specific behavioral match.** Higher bar than step 3. The body must describe the *specific behavior* the test checks — not just the general feature area. If the connection is only that they're in the same feature area, that is not enough.

For example: a test that checks `ValidateCardNumber()` rejects expired cards does NOT match an issue about "Payment processing" whose body discusses order workflows. It DOES match an issue whose body says "add validation for expired card numbers," even without naming `ValidateCardNumber()`.

**5. Decide.**
- **Tag** if you found a strong match in step 3, or a specific behavioral match in step 4.
- **Skip** if potential issues are about unrelated functionality.
- **When uncertain**, lean slightly toward skipping. An untagged test can be found later; an incorrect tag must be identified and removed during review.

**6. Check for duplicates.** The `Issues` column shows what's already tagged. Do not re-add existing tags.

### Common errors to avoid

Do NOT match on superficial keyword overlap or feature-area proximity:

- **Keyword overlap in title only**: Test mentions "payments" → matching any issue with "payments" in the title when the body describes different work
- **Feature-area match without behavioral match**: Test checks `FormatReceipt()` handles refunds → matching an issue about "Payment processing" whose body discusses order validation
- **Body contradicts title**: Issue title sounds relevant but the body describes unrelated work — always trust the body over the title
- **Incidental file changes**: Issue is about infrastructure or cleanup that happened to touch the same file
- **Description vs. code mismatch**: Matching on the test description when `TestCode` shows the test checks something different

### Special cases

**Tests with `intIssue` in `ExpectUserAccepts()` calls**: Tag with the referenced issue number unless the value is obviously test data (e.g., 1, 12, 123).

**Tests already tagged**: If `Issues` is non-empty and looks correct, skip. When adding new tags, preserve existing ones and keep issue numbers in ascending order.

**Disambiguating similar issues**: Read each body completely and pick the one that most specifically describes the behavior tested. Reject issues about infrastructure or the broader feature area. Tag with all if multiple genuinely pass steps 3–4, but this should be uncommon.

### Example

```r
test <- dfTestIssueContext[5, ]

test$Test
# "ProcessPayment handles declined cards"
test$TestCode[[1]]
# test_that("ProcessPayment handles declined cards", {
#   local_mocked_bindings(GetAPIResult = function(...) list(result = "declined"))
#   expect_error(ProcessPayment(), class = "error-payment_declined")
# })

test$PotentialIssueDetails[[1]]
#   Issue Title                                Body
#   42    "Payment system overhaul"            "Refactor payment module architecture..."
#   87    "Payment processing error handling"  "Implement ProcessPayment() to handle declined
#                                               cards, expired cards, and..."
#   104   "Add logging to payment module"      "Add debug logging throughout payment..."

# Issue 42: about refactoring, not implementing ProcessPayment → NO
# Issue 87: explicitly mentions ProcessPayment() and declined cards → YES
# Issue 104: about logging, unrelated → NO

# Decision: Tag with #87
```

## Step 3: Tag untagged tests by title match

After commit-based matching, attempt to tag any tests that still have no issue tags. Fetch the full issue list once before the file loop begins:

```r
dfAllIssues <- FetchRepoIssues()  # returns Issue (integer) and Title (character) at minimum
```

For each test where `length(test$Issues[[1]]) == 0`, compare the test description to all issue titles. Apply a **higher bar** than Step 2 — titles alone carry less signal than commit history plus issue bodies. Only tag if the title unambiguously describes the specific behavior the test checks. Do not match on shared keywords or feature-area overlap.

Mark these in `test_tag_reasons.qmd` with **"Title match"** in the Reason column so reviewers know they are lower-confidence.

## Step 4: Edit test files

### Tag format

- Single issue: `test_that("does something (#123)", { ... })`
- Multiple issues: `test_that("does something (#123, #456)", { ... })`

### Editing guidelines

- **Only edit** the parenthetical issue tags in the `test_that()` description. Do not change anything else.
- Preserve existing tags; sort issue numbers ascending.
- Use `File`, `LineStart`, and `LineEnd` to locate each test.
- Batch edits to the same file; preserve code style and indentation.

### Editing code

```r
IssuesToTag <- 87L  # include existing tags if any
test$IssueTags <- glue::glue("#{IssuesToTag}") |>
  glue::glue_collapse(sep = ", ")

readLines(test$File) |>
  stringr::str_replace(
    stringr::fixed(test$Test),
    glue::glue_data(test, "{Test} ({IssueTags})")
  ) |>
  writeLines(test$File)
```

## Step 5: Log reasoning

Keep a running log in `test_tag_reasons.qmd` (renders to HTML). Include a table per file with columns "Test", "Issue" (as a GitHub link), and "Reason" (1–2 sentences). Update as you work, not all at once at the end.

## Validation

After tagging:
- Run `devtools::test(reporter = "check")` — only descriptions should change (snapshots may update)
- Re-run `ExtractTestsFromFiles()` and confirm `Issues` contains the tagged numbers
- Render `test_tag_reasons.qmd`

---

This was the prompt (basically adding a bit onto the skill that I thought of but didn't incorporate into the skill; the results of the bonus round were spotty, so I don't plan to incorporate that into the skill):

> tag tests with issues. After you're done with the skill, check for closed (completed) issues in this repo that aren't tagged to any tests, and try to tag at least one test with each issue when it makes sense to do so (eg, don't worry about finding a test to map to issue #107 "Implement gsm.utils v0.2.0 workflows" because that issue purely relates to the @.github/workflows/ folder, not anything in this package itself, but DO try to find a test for #89 "Add site risk score metric to reportingResults" even if the first pass fails to do so). Include this final round of tagging in the `test_tag_reasons.qmd` report mentioned in the skill.

**Comment by jonthegeek** [comment]
Converting to draft until qcthat release to minimize the chance that this gets released accidentally. I'll clean out unused AI-related files at that time as well.

**Comment by jonthegeek** [comment]
@lauramaxwell After merging in the current version of dev, the qualification tests are still throwing warnings:

```
Not all columns of Raw_AE in the spec are in the expected format, improperly formatted columns are: aetoxgr
Not all columns of Raw_SUBJ in the spec are in the expected format, improperly formatted columns are: timeontreatment
Not all columns of Raw_SUBJ in the spec are in the expected format, improperly formatted columns are: agerep
```

It looks like we need a gsm.mapping release to fix those if we want to get rid of these warnings (which currently freak out qcthat).

Separately I'll make qcthat configurable around this, but ideally we really shouldn't allow random warnings through like this. I did a temporary update to suppress that warning just to make sure that's the only thing failing.

**Review by lauramaxwell — COMMENTED** [review]
making a comment prior to next qcthat run

File: tests/testthat/_snaps/util-checkSpec.md (hunk 7 above)
Comment (nandriychuk) [reply tests/testthat/_snaps/util-checkSpec.md]: I don't think we expect snapshots to have issues numbers? If so, should we update a skill to exclude these from matching?

**Review by lauramaxwell — APPROVED** [review]
Made a bunch of edits- mostly on the issues themselves, but tagged 3 additional issues with tests for `Flag()` and `Flag_Accrual()`.

This is looking like a good start to me, but happy to discuss any of the edits I've made to get the report to be focused on features and unit-test-able issues.

**Review by nandriychuk — APPROVED** [review]
